ui/plugins/auto_classification/controller.js

```diff
@@ -1,372 +1,924 @@
+            if (!src.hasOwnProperty(key)) {
+                continue;
+            }
+            var descriptor = Object.getOwnPropertyDescriptor(src, key);
+            if (descriptor && descriptor.get) {
+                Object.defineProperty(dest, key,
+                                      {get: descriptor.get,
+                                       set: descriptor.set,
+                                       enumerable: descriptor.enumerable,
+                                       configurable: descriptor.configurable});
+            } else {
+                dest[key] = src[key];
+            }
+        }
+    }
+    return dest;
+};
+
+var isHelpfulLine = function(lineData) {
+    lineData = lineData.replace(/\s*(.*)\s*/, "$1");
+
+    var blacklist = {
+        'automation.py': true,
+        'remoteautomation.py': true,
+        'Shutdown': true,
+        'undefined': true,
+        'Main app process exited normally': true,
+        'Traceback (most recent call last):': true,
+        'Return code: 0': true,
+        'Return code: 1': true,
+        'Return code: 2': true,
+        'Return code: 9': true,
+        'Return code: 10': true,
+        'Exiting 1': true,
+        'Exiting 9': true,
+        'CrashingThread(void *)': true,
+        'libSystem.B.dylib + 0xd7a': true,
+        'linux-gate.so + 0x424': true,
+        'TypeError: content is null': true,
+        'leakcheck': true
+    };
+
+    return lineData.length > 4 && !blacklist.hasOwnProperty(lineData);
+};
+
+treeherder.factory('ThClassificationOption', [
+    function() {
+        var ThClassificationOption = function(type, id, bugNumber, bugSummary, matches) {
+            extendProperties(this, {
+                type: type,
+                id: id,
+                bugNumber: bugNumber || null,
+                bugSummary: bugSummary || null,
+                hasBug: bugNumber ? true : false, // Did the option have a bug when first created
+                matches: matches || null,
+                always: false, // For type = 'ignore' ignore just for this job or
+                               // also for future cases where the same line is matched
+                isBest: false,
+                get icon() {
+                    return this.isBest ? "autoclassified" : 'none';
+                }
+            });
+        };
+        return ThClassificationOption;
+    }
+]);
 
-        $log.debug("error classification plugin initialized");
+treeherder.factory('ThStructuredLinePersist', ['$q',
+                                               'thNotify',
+                                               'thTabs',
+                                               'ThFailureLinesModel',
+                                               'ThClassifiedFailuresModel',
+    function($q, thNotify, thTabs, ThFailureLinesModel, ThClassifiedFailuresModel) {
+        /*
+         When saving a structured line, we need to accound for the following cases:
```

Comment:
How DARE you spell "account" with a D?!?  I... what?  
How DARE you accuse me of overreacting to something trivial?!?  
Oh...  that was just a... typo?  Hrm...  where are my meds...